nexus/db-queries/src/db/datastore/dns.rs

```diff
@@ -0,0 +1,1005 @@
+    /// number being read
+    #[cfg(test)]
+    pub async fn dns_config_read_version_test(
+        &self,
+        opctx: &OpContext,
+        log: &slog::Logger,
+        batch_size: NonZeroU32,
+        version: &DnsVersion,
+    ) -> Result<DnsConfigParams, Error> {
+        self.dns_config_read_version(opctx, log, batch_size, version).await
+    }
+
+    /// Private helper for reading a specific version of a group's DNS config
+    async fn dns_config_read_version(
+        &self,
+        opctx: &OpContext,
+        log: &slog::Logger,
+        batch_size: NonZeroU32,
+        version: &DnsVersion,
+    ) -> Result<DnsConfigParams, Error> {
+        debug!(log, "reading DNS config");
+        debug!(log, "reading DNS zones");
+        let dns_group = version.dns_group;
+        let dns_zones = self
+            .dns_zones_list(
+                opctx,
+                dns_group,
+                &DataPageParams {
+                    marker: None,
+                    direction: dropshot::PaginationOrder::Ascending,
+                    limit: NonZeroU32::try_from(NMAX_DNS_ZONES).unwrap(),
+                },
+            )
+            .await
+            .with_internal_context(|| {
+                format!("listing internal zones for DNS group {:?}", dns_group)
+            })?;
+        debug!(log, "found zones"; "count" => dns_zones.len());
+
+        bail_unless!(
+            dns_zones.len() < usize::try_from(NMAX_DNS_ZONES).unwrap()
+        );
+
+        let mut zones = Vec::with_capacity(dns_zones.len());
+        for zone in dns_zones {
+            let mut total_found = 0;
+            let mut zone_records = Vec::new();
+            let mut marker = None;
+
+            loop {
+                debug!(log, "listing DNS names for zone";
+                    "dns_zone_id" => zone.id.to_string(),
+                    "dns_zone_name" => &zone.zone_name,
+                    "version" => i64::from(&version.version.0),
+                    "found_so_far" => total_found,
+                    "batch_size" => batch_size.get(),
+                );
+                let pagparams = DataPageParams {
+                    marker: marker.as_ref(),
+                    direction: dropshot::PaginationOrder::Ascending,
+                    limit: batch_size,
+                };
+                let names_batch = self
+                    .dns_names_list(opctx, zone.id, version.version, &pagparams)
+                    .await?;
+                let nfound = names_batch.len();
+                if nfound == 0 {
+                    break;
+                }
+
+                total_found += nfound;
+                let last = &names_batch[nfound - 1];
+                marker = Some(last.key.name.clone());
+
+                zone_records.extend(names_batch.into_iter());
+
+                if nfound < usize::try_from(batch_size.get()).unwrap() {
+                    break;
+                }
```

Comment:
Nit: I think `total_found` is always equal to `zone_records.len()`, right?

This might make the control flow marginally simpler:

```suggestion
                let done = names_batch.len() < usize::try_from(batch_size.get()).unwrap();
                if let Some(last) = names_batch.last() {
                  marker = Some(last.key.name.clone());
                }
                zone_records.extend(names_batch.into_iter());
                if done {
                    break;
                }
```